docs/source/architecture.rst

```diff
@@ -1,19 +1,740 @@
+
+.. _heavy_material:
+
+Heavy material nodes
+++++++++++++++++++++
+
+Heavy material nodes are stateful and require recovery and content revalidation (proof-of-storage), both periodically and upon rejoining the network. On the Insolar network, heavy material nodes do the following:
+
+* provide long-term data storage and scalability of storage capacity;
+* store all data received from :ref:`light material nodes <light_material>` (and, in turn, from :ref:`virtual nodes <virtual>`);
+* check data integrity but are unable to introduce or change data or form a block;
+* ensure the required level of block replication and the maximum data density (scattering) to reduce the impact of data leakage from a single material node (heavy or light).
+
+Heavy material nodes differ significantly from other nodes -- they store lots of data and must take additional measures to mitigate the following risks:
+
+* losing (or corrupting) data but not having enough copies, or
+* data leakage caused by the accumulation of too much data on a single node.
+
+Heavy material node's implementation is simplified for the TestNet 1.1 and will gradually extend during the development of Insolar's enterprise version.
+
+Moreover, additional network protocol is implemented to maintain backups and archival storage nodes without burdening the main Insolar network consensus.
+
+.. _dynamic_roles:
+
+Dynamic Roles
+^^^^^^^^^^^^^
+
+In addition to the node's static role, it can be equipped with dynamic ones -- roles able to change.
+
+:ref:`Virtual nodes <virtual>` can have the following roles and respective responsibilities:
+
+* **Virtual executor** handles operations on a :term:`lifeline` and builds new :term:`object <object>` states.
+* **Virtual validator** verifies virtual executor's actions from previous :term:`pulses <pulse>`.
+
+:ref:`Light material nodes <light_material>` can have the following roles and respective responsibilities:
+
+* **Material executor** forms new :term:`blocks <jet drop>` and grants access to previous blocks.
+* **Material validator** checks the block's validity and consistency.
+* **Material stash** caches hot data and relevant indices (current states of all :term:`objects <object>`) and syncs the indices among other stash nodes.
+
+In essence, all the nodes take part in two kinds of :ref:`execution and validation <execution_validation>` procedures, depending on their dynamic roles: **virtual** and **material**. :ref:`Heavy material nodes <heavy_material>` rely on validation performed by light material ones.
+
+A node can have multiple dynamic roles, e.g., a virtual node can be selected via the :term:`entropy <pulse>` to be an executor for one :term:`lifeline` and a validator of another.
+
```

Comment:
There are also HM Executors and HM Stash. HMS are all non-executor nodes. And HME role is election-based compared to other modes.